### PR TITLE
EI-704 Add new Meta service and SparqlQuery RPC call

### PR DESCRIPTION
**Pull request by vtermanis**
**Note**: The initial implementation will only target "synchronous" search (i.e. without independent dispatch & listen-for-results RPC calls)

### CODE REVIEW

File: proto/iotics/api/meta.proto [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    uint64 seqNum = 2;
+
+    // Indicates whether this is the last chunk from a given host, for a specific request. Results for different
+    // requests can be identified by setting a unique clientRef in the request headers.
+    bool last = 3;
+
+    // Result error status (only applicable to local results, i.e. when remoteHostId is unset). If set, this will
+    // indicate a problem with running the query (e.g. invalid syntax) as opposed to a more general issue (in which case
+    // the standard gRPC error mechanism will be used and the stream terminated).
+    google.rpc.Status status = 4;
```
Comment (vtermanis) [inline proto/iotics/api/meta.proto:99]: See [Slack discussion](https://iotic-labs.slack.com/archives/C02A46LSNSC/p1643389916141779) regarding when status should be set.
I would argue that it only makes sense to set/return it for a local response. If something goes wrong remotely, we send back nothing.

File: proto/iotics/api/meta.proto [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    // The actual result content type (as raw mime type). For the hinted type (in a request), typically the returned
+    // type can be:
+    // - SPARQL_JSON - application/sparql-results+json
```
Comment (vtermanis) [inline proto/iotics/api/meta.proto]: Maybe instead of listing these, one can just say: "_Refer to the standards specifications of SparqlResultType for relevant mime types_"

**Edit**: If we verify the return type, then could just return the request enum instead.

File: proto/iotics/api/meta.proto [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+// MetaAPI enables querying of metadata associated with Twins and Feeds.
+service MetaAPI {
+
+  // SparqlQuery performs a SPARQL 1.1 query and returns one or more results, each as a sequence of chunks. Note that:
+  // - Chunks for a particular result will arrive in-order though they might be interleaved with chunks from other
+  //   results (when performing a non-local query). See scope parameter in SparqlQueryRequest.
+  // - The call will only complete once the (specified or host default) request timeout has been reached. The client can
+  //   choose to end the stream early once they have received enough results. (E.g. in the case of Scope.LOCAL this
+  //   would be after the one and only sequence of chunks has been received.)
+  rpc SparqlQuery(SparqlQueryRequest) returns (stream SparqlQueryResultChunk) {}
```
Comment (StephanieBracaloni) [inline proto/iotics/api/meta.proto]: In most of the other services we are going with Request/Response pattern (not Result)
I'm not sure "chunk" needs to be in the response name

File: proto/iotics/api/meta.proto [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+// SparqlQueryRequest describes a SPARQL query.
+message SparqlQueryRequest {
+
+  // SPARQL request payload.
+  message Payload {
+
+    // The desired result content type. Note that the actual content (mime) type could be different and will be returned
+    // in result chunks.
+    SparqlResultType resultTypeHint = 1;
```
Comment (StephanieBracaloni) [inline proto/iotics/api/meta.proto]: resultContentType ?

File: proto/iotics/api/meta.proto (hunk 3 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: Originally (and in PoC) I did have response.. but:
The reason I thought of result.. is that the response is made up of many results which each might be made up of many chunks.
Typically for querying - you get a result, not a response. 

But I get that for RPCs one should to "Request" & "Response"

.. to be fair `SparqlQueryRequest` itself sounds a bit wordy - but I was thinking we might also have later:
- `DispatchSparqlQuery` (for the async part)
- `SparqlUpdateRequest` (if we do indeed expose local write down the line)

Comment (vtermanis) [reply proto/iotics/api/meta.proto]: > I'm not sure "chunk" needs to be in the response name

If it's not in there, to me it's not clear that individual responses/result (whatever the name) are in fact only partial (when the result is larger).
Having to say "You will get multiple responses.. each containing a chunk of the response" - sounds not as good as "You will get multiple chunks which make up the full response/result"

File: proto/iotics/api/meta.proto (hunk 4 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: I had exactly that before - I was wondering whether _hint_ is more appropriate because this is after all an `Accept:` and as per note against the enum, it's not guaranteed to that.
However, `resultContentType` would also be OK.

**Review by mihai-dinculescu — APPROVED** [review]
LGTM

File: proto/iotics/api/meta.proto [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  // - Chunks for a particular result will arrive in-order though they might be interleaved with chunks from other
+  //   results (when performing a non-local query). See scope parameter in SparqlQueryRequest.
+  // - The call will only complete once the (specified or host default) request timeout has been reached. The client can
+  //   choose to end the stream early once they have received enough results. (E.g. in the case of Scope.LOCAL this
+  //   would be after the one and only sequence of chunks has been received.)
+  rpc SparqlQuery(SparqlQueryRequest) returns (stream SparqlQueryResultChunk) {}
+}
+
+// Available result content types for SPARQL requests. Note that applicable content types depend on the type of query.
+// Specifying an incompatible result type can lead to the result being returned with a different content type.
```
Comment (mihai-dinculescu) [inline proto/iotics/api/meta.proto]: Can we have an error returned if the requested content type can't be honoured? This would eliminate a lot of surprises.

Comment (vtermanis) [reply proto/iotics/api/meta.proto]: > Can we have an error returned if the requested content type can't be honoured?

We discussed exactly this (and the [PoC I wrote returned an error in this case](https://github.com/Iotic-Labs/iotic-service-eventbus/blob/ei-605/pkg/meta/sparql.go#L289-L300) - see also "things to bear in mind" [here](https://ioticlabs.atlassian.net/browse/EI-707))

However: If you run a SPARQL query yourself (directly) - the server won't tell you "it's wrong" **or** a different implementation (other than Fuseki) might respond in different way. (E.g. Fuseki will just return `rdf-xml`)

Comment (mihai-dinculescu) [reply proto/iotics/api/meta.proto]: Do we have to do the same as Fuseki?
The client requests a specific format and they write code that works with the format they expect. They need to know ASAP that the format they have in mind is not suitable.
Pushing this check to the client side is further complicated by the fact that the response content type is set on each chunk, so they would have to check each and every chunk to make sure that there is no funny business going on.

File: proto/iotics/api/meta.proto (hunk 3 above)
Comment (StephanieBracaloni) [reply proto/iotics/api/meta.proto]: I'm not 100% (Ican not check now, I have to drop off ) but I thing Request/Response is the pattern used in the prototool guidelines https://github.com/uber/prototool/blob/dev/docs/lint.md

File: proto/iotics/api/meta.proto (hunk 5 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: > Do we have to do the same as Fuseki?

Not at all - is not what I was suggesting - I was just highlighting what one implementation does when mime type is unsuitable

> They need to know ASAP that the format they have in mind is not suitable.

During local refinement for the implementation (with @StephanieBracaloni & @conbon) it was decided to defer that handling to the client.
However, I've made a mental note of this and we can most definitely offer that. I agree with you it's nice for the user to be notified that they won't get what they expect. The 2nd upside is that we don't need to include the "actual" mime type - we can just re-state the chosen (enum) type.

File: proto/iotics/api/meta.proto (hunk 3 above)
Comment (conbon) [reply proto/iotics/api/meta.proto]: REQUEST_RESPONSE_NAMES_MATCH_RPC - "Verifies that all request names are of the pattern RpcNameRequest and all response names are of the pattern RpcNameResponse."

I guess the docs just need to be clear about the kind of chunked response returned.

File: proto/iotics/api/meta.proto (hunk 4 above)
Comment (conbon) [reply proto/iotics/api/meta.proto]: definitely want to update this if thinking of failing fast server side on an incorrect type/query mapping

File: proto/iotics/api/meta.proto [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+// Note that chunks are guaranteed to arrive in-order (for a specific result from a host) but could be duplicated. It is
+// up to the client to ignore duplicates (based on sequence number).
+message SparqlQueryResultChunk {
+
+  // Payload of the query result chunk
+  message Payload {
+
+    // Result host identifier. Indicates from which host this result chunk came from. For a local result, this field
+    // will be unset.
+    HostID remoteHostId = 1;
```
Comment (conbon) [inline proto/iotics/api/meta.proto:87]: Not sure I like the reliance on absence of value to indicate a local host value.

Seems like this might simplify code in multiple places to just always set the value and construct responses in a similar fashion for local/remote.

File: proto/iotics/api/meta.proto [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    // The actual result content type (as raw mime type). For the hinted type (in a request), typically the returned
+    // type can be:
+    // - SPARQL_JSON - application/sparql-results+json
+    // - SPARQL_XML - application/sparql-results+xml
+    // - SPARQL_CSV - text/csv
+    // - RDF_TURTLE - text/turtle
+    // - RDF_XML - application/rdf+xml
+    // - RDF_NTRIPLES - application/n-triples
+    // - RDF_NQUADS - application/n-quads
+    string actualType = 5;
```
Comment (conbon) [inline proto/iotics/api/meta.proto]: this may also need updated as per above if choosing to fail fast and not swap out types under the user. - contentType

File: proto/iotics/api/meta.proto [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+    // Query result chunk, encoded according to actualType. The maximum size of each chunk is host-specific.
```
Comment (conbon) [inline proto/iotics/api/meta.proto:104]: if changing actualType, update comment here.

Comment (conbon) [reply proto/iotics/api/meta.proto:104]: "The maximum size of each chunk is host-specific." - Does this mean a rogue host could DOS the network with massive chunk settings? Is there an absolute maximum?

Comment (conbon) [reply proto/iotics/api/meta.proto:104]: Should this be a value (chunkSize) specified in the request, up until the max value set by host?

File: proto/iotics/api/meta.proto [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    bytes resultChunk = 6;
+  }
+
+  // Headers fo the query result. clientRef within can be used to identify which query the result applies to.
```
Comment (conbon) [inline proto/iotics/api/meta.proto]: *for

**Review by conbon — COMMENTED** [review]
Some comments

File: proto/iotics/api/meta.proto (hunk 3 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: Thanks, I'm **well** aware of [the linter check](https://github.com/uber/prototool/blob/a6d064684c011c8482f4218c681a66a64031fb1d/internal/lint/check_request_response_names_match_rpc.go#L48-L56) and I know that the rest of the API uses that - so of course this one will too.

Am I right in thinking that we don't actually run the uber linter at the moment? (It would be great if we could.) Is there a reason for that? (I mean - this PR and the spec-qapi one pass, despite the lack of "response" in the response type right now.)

File: proto/iotics/api/meta.proto (hunk 6 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto:87]: > to just always set the value

Can you explain what you mean? Do you mean set it to the local host ID?

The suggested behaviour is the same as for search currently (and also discover/follow - where absence/nil value means it's a local twin/feed).

The client is not aware of the local host ID (without performing additional lookups) - so if it were set - they'd have to either hard-code it locally first or interrogate to determine whether it's local.

File: proto/iotics/api/meta.proto (hunk 7 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: @conbon - I'm well aware of that :+1:   - this would just be the enum

Comment (conbon) [reply proto/iotics/api/meta.proto]: nice one.

File: proto/iotics/api/meta.proto (hunk 8 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto:104]: > Should this be a value (chunkSize) specified in the request, up until the max value set by host?

We could offer that - but then we'd also definitely want a minimum value (sending tiny chunks is also a type of DoS - one has to send many more messages/packets/whatever than with larger ones).
So maybe it should be a chunk size hint. Note that a remote host might choose to respond with a differently sized chunk. This would mean that locally one might have to split (or combine) chunks based on the client's suggestion.
Maybe a "max chunk size" client option might be best: "_I can handle anything up to at most this size_"

> Does this mean a rogue host could DOS the network with massive chunk settings? Is there an absolute maximum?

Chunks apply to responses, so that would mean a remote host would send massive chunks to the originating (local) host (i.e. a point-to-point connection). We have the option (not implemented yet) to reject oversized payloads from the p2p network - that would be a general issue - not specific to this type of request. It's a good one to look into but I think that is a general "hardening the host" issue.
@smartrics  - do you already have something in the road-map about making hosts robust when communicating with other peers/hosts?

File: proto/iotics/api/meta.proto (hunk 4 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: FYI - "failing fast" will still mean running the query until response headers are received from SPARQL endpoint and `Accept` can be compared to `Content-Type`.

File: proto/iotics/api/meta.proto (hunk 8 above)
Comment (conbon) [reply proto/iotics/api/meta.proto:104]: Yes, I agree it's definitely a more general concern than just this request.

File: proto/iotics/api/meta.proto (hunk 3 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: e98aee517d91177a874f70b8cdec0821d66a7c75

File: proto/iotics/api/meta.proto (hunk 2 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: e98aee517d91177a874f70b8cdec0821d66a7c75

Removed this due to performing type check

File: proto/iotics/api/meta.proto (hunk 4 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: e98aee517d91177a874f70b8cdec0821d66a7c75

File: proto/iotics/api/meta.proto (hunk 5 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: @mihai-dinculescu  - plan is to verify the content type - i.e. it will be guaranteed (or will get error status for local response)
e98aee517d91177a874f70b8cdec0821d66a7c75

File: proto/iotics/api/meta.proto (hunk 7 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: e98aee517d91177a874f70b8cdec0821d66a7c75

File: proto/iotics/api/meta.proto (hunk 9 above)
Comment (vtermanis) [reply proto/iotics/api/meta.proto]: Good spot

e98aee517d91177a874f70b8cdec0821d66a7c75

File: proto/iotics/api/meta.proto [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  // Applicable to SELECT/ASK (SPARQL 1.1 Query Results XML Format)
+  SPARQL_XML = 1;
+  // Applicable to SELECT/ASK (SPARQL 1.1. Query Results CSV Format)
+  SPARQL_CSV = 2;
+  // Applicable to CONSTRUCT/DESCRIBE (Terse RDF Triple Language)
+  RDF_TURTLE = 3;
+  // Applicable to CONSTRUCT/DESCRIBE (RDF 1.1 XML)
+  RDF_XML = 4;
+  // Applicable to CONSTRUCT/DESCRIBE (RDF 1.1 N-Triples)
+  RDF_NTRIPLES = 5;
```
Comment (vtermanis) [inline proto/iotics/api/meta.proto:49]: FYI - removed N-Quads since CONSTRUCT & DESCRIBE refer to a single graph. Quads are only applicable to the graph store protocol, when retrieving more than one graph.

File: proto/iotics/api/meta.proto (hunk 6 above)
Comment (StephanieBracaloni) [reply proto/iotics/api/meta.proto:87]: That's something that has been raised several times about the search. It will be easier to make something consistent when we will replace the `remoteHostId`  (peerID) with the host twin DID. We don't necessarily need to address that right now since the work to return the hosttwinDID is not trivial but it is great to keep that in mind

File: proto/iotics/api/meta.proto (hunk 1 above)
Comment (StephanieBracaloni) [reply proto/iotics/api/meta.proto:99]: :+1: 
https://iotic-labs.slack.com/archives/C02A46LSNSC/p1643962003610629?thread_ts=1643389916.141779&cid=C02A46LSNSC

File: proto/iotics/api/meta.proto (hunk 8 above)
Comment (StephanieBracaloni) [reply proto/iotics/api/meta.proto:104]: No on the roadmap yet but we will have to consider Network API security as soon as we will offer on premise deployment
- host twin authenticity 
- host banishment
- signed payload
- ....

Comment (StephanieBracaloni) [reply proto/iotics/api/meta.proto:104]: > This would mean that locally one might have to split (or combine) chunks based on the client's suggestion.

The client might already expect different chunk sizes since it is configurable by host? :eyes:

Comment (vtermanis) [reply proto/iotics/api/meta.proto:104]: > The client might already expect different chunk sizes since it is configurable by host?

@StephanieBracaloni I don't understand what you're saying. Do you mean: Chunk size should be configurable in host, **not** per-request? (The original proposal.)

Note that grpc ["max receive message size"](https://github.com/grpc/grpc-go/blob/master/rpc_util.go#L572-L573) is a client-only option. I'm not sure we should let the client set an expectation. If it receives something too large it doesn't want to handle, it can ignore it. (Otherwise we're adding more logic / negotiation )

Maximum size is something that also applies to:
- search results (if had a higher limit than now)
- twin/feed listing (again, if configured higher limit)
- feed share receive

So it feels like there should be a general solution (at least in that we consider it as such) first before going down the route of request-specific flags.

Comment (StephanieBracaloni) [reply proto/iotics/api/meta.proto:104]: I was thinking about a chunk size per host for the first approach.
Then we can also offer the option to the end-user if needed (with a max value in the host).

In both cases, the end-user should expect to receive different chunk sizes since the remote request will come from a remote host with a different configuration.

File: proto/iotics/api/meta.proto [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  // SPARQL request payload
+  Payload payload = 3;
+}
+
+// SparqlQueryResponse is a part of a result for a SPARQL query request. Multiple chunks form a complete result. Related
+// chunks can be identified by a combination of:
+// - The host ID (unset for local results)
+// - Client reference (in headers, set by caller)
+// - Chunk sequence number
+// Note that chunks are guaranteed to arrive in-order (for a specific result from a host) but could be duplicated. It is
```
Comment (StephanieBracaloni) [inline proto/iotics/api/meta.proto]: I don't think we can guarantee the order, especially since we will have to handle chunks over the network. The end-user should be able to reorder the messages thanks to the seq number.

**Review by StephanieBracaloni — CHANGES_REQUESTED** [review]

File: proto/iotics/api/meta.proto (hunk 11 above)
Comment (StephanieBracaloni) [reply proto/iotics/api/meta.proto]: Is libp2p guaranteeing the message order?
I would go with not going with this level of details there until we have internal feedback about the feature

Comment (vtermanis) [reply proto/iotics/api/meta.proto]: I've removed the note about in-order arrival

**Review by StephanieBracaloni — APPROVED** [review]
LGTM
